### PR TITLE
rvv optimazation for AbsVal and ReLU

### PR DESCRIPTION
**Pull request by thelastlin**

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=3001) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=3001) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=3001) it.</sub>

**Comment by nihui** [comment]
please fix https://github.com/Tencent/ncnn/runs/2808997547 test failure

**Comment by thelastlin** [comment]
> 
> 
> please fix https://github.com/Tencent/ncnn/runs/2808997547 test failure

now it should be fixed in https://github.com/Tencent/ncnn/pull/3001/commits/f06518d1a9ac630cd15535be680cf78b0acd4ade

File: src/layer/riscv/relu_riscv.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+protected:
+#if __riscv_vector && __riscv_zfh
+    int forward_inplace_fp16s(Mat& bottom_top_blob, const Option& opt) const;
+    int forward_inplace_fp16sa(Mat& bottom_top_blob, const Option& opt) const;
+#endif
+};
+
+} // namespace ncnn
+
+#endif // LAYER_RELU_ARM_H
```
Comment (nihui) [inline src/layer/riscv/relu_riscv.h:38]: LAYER_RELU_RISCV_H

File: src/layer/riscv/absval_riscv.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+protected:
+#if __riscv_vector && __riscv_zfh
+    int forward_inplace_fp16s(Mat& bottom_top_blob, const Option& opt) const;
+    int forward_inplace_fp16sa(Mat& bottom_top_blob, const Option& opt) const;
+#endif
+};
+
+} // namespace ncnn
+
+#endif // LAYER_ABSVAL_ARM_H
```
Comment (nihui) [inline src/layer/riscv/absval_riscv.h:38]: LAYER_ABSVAL_RISCV_H

File: src/layer/riscv/absval_riscv.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+#include "absval_riscv.h"
+
+#if __riscv_vector
+#ifdef RVV_SPEC_0_7
+#include "riscv_v_071_fix.h"
+#else
+#include <riscv_vector.h>
+#endif
+#include "rvv_mathfun.h"
+#include "rvv_mathfun_fp16s.h"
```
Comment (nihui) [inline src/layer/riscv/absval_riscv.cpp:24]: these two headers are not needed

File: src/layer/riscv/relu_riscv.cpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+#include "relu_riscv.h"
+
+#if __riscv_vector
+#ifdef RVV_SPEC_0_7
+#include "riscv_v_071_fix.h"
+#else
+#include <riscv_vector.h>
+#endif
+#include "rvv_mathfun.h"
+#include "rvv_mathfun_fp16s.h"
```
Comment (nihui) [inline src/layer/riscv/relu_riscv.cpp:24]: these two headers are not needed

File: src/layer/riscv/absval_riscv.cpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    int elembits = bottom_top_blob.elembits();
+
+#if __riscv_vector && __riscv_zfh
+    if (opt.use_fp16_storage && elembits == 16)
+    {
+        if (opt.use_fp16_arithmetic)
+            return forward_inplace_fp16sa(bottom_top_blob, opt);
+        else
+            return forward_inplace_fp16s(bottom_top_blob, opt);
+    }
```
Comment (nihui) [inline src/layer/riscv/absval_riscv.cpp:59]: For operators that has effectively precision difference caused by the intermediate accumulator type, we implement the two variants, fp16 and fp16sa.
For abs and relu, it is fine to implement only one fp16s function that uses fp16 intermediate type.

File: src/layer/riscv/relu_riscv.cpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+{
+    int elembits = bottom_top_blob.elembits();
+
+#if __riscv_vector && __riscv_zfh
+    if (opt.use_fp16_storage && elembits == 16)
+    {
+        if (opt.use_fp16_arithmetic)
+            return forward_inplace_fp16sa(bottom_top_blob, opt);
+        else
+            return forward_inplace_fp16s(bottom_top_blob, opt);
```
Comment (nihui) [inline src/layer/riscv/relu_riscv.cpp:49]: see related comment in absval.cpp  :)

**Review by nihui — CHANGES_REQUESTED** [review]

**Review by nihui — APPROVED** [review]

**Comment by nihui** [comment]
Thanks for your contribution !

**Pull request merged**